### PR TITLE
Config Yaml changes and default values

### PR DESCRIPTION
**Pull request by paarthneekhara**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/tts/speechllm/conf/megatron_t5_speechllm_inference_model.yaml [hunk 1]
```diff
@@ -54,8 +54,6 @@ model:
   validation_metric: accuracy
   num_speech_tokens: 10112 # Vocabulary size pertaining to speech
   seq_pattern: "parallel" # parallel, delay_parallel, flatten
-  speech_head_type: "linear" # token_level, linear
-  cross_entropy_type: "vocab_parallel" # regular, vocab_parallel
   temperature: 0.7 # Temperature to be used for inference
   top_k: 80 # Top k to be used for inference
```
Comment (blisc) [inline examples/tts/speechllm/conf/megatron_t5_speechllm_inference_model.yaml]: In the code, the defaults are `temperature = self.cfg.get('temperature', 0.85)  # Set temp 0.01 for greedy decoding`, should we update the configs to match?

File: examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml [hunk 2]
```diff
@@ -63,7 +63,12 @@ model:
   save_nemo_on_validation_end: True # Saves an inference ready .nemo file every time a checkpoint is saved during training.
   existing_tasks: []
   new_tasks: ["squad"]
-
+  codecmodel_type: nemo_codec
+  codecmodel_path: ???
+  english_only_model: true
+  context_conditioning: decoder
+  use_flash_attention: false
```
Comment (blisc) [inline examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml:70]: This is true in the other inference file, we should agree on one

File: examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml [hunk 3]
```diff
@@ -100,6 +105,10 @@ model:
     speech_offset: 30000
     train_task: asr
```
Comment (blisc) [inline examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml]: Should we update the task on both inference yamls?

File: examples/tts/speechllm/conf/megatron_t5_speechllm_inference_model.yaml (hunk 1 above)
Comment (paarthneekhara) [reply examples/tts/speechllm/conf/megatron_t5_speechllm_inference_model.yaml]: Yes, makes sense. Let me do that

File: examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml (hunk 2 above)
Comment (paarthneekhara) [reply examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml:70]: Sure, let's leave it as false? (Since pretrained model and interspeech checkpoitns are all false I believe)

File: examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml (hunk 3 above)
Comment (paarthneekhara) [reply examples/tts/speechllm/conf/megatron_t5_speechllm_inference.yaml]: Yes, my bad. Let me update this.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml [hunk 4]
```diff
@@ -71,7 +69,7 @@ model:
   context_conditioning: decoder
   train_from_scratch: true
   override_tokenizer_vocab_file: ???
-  use_flash_attention: true
+  use_flash_attention: false
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml:72]: hmm...do you suggest to enable `use_flash_attention` during training? I did not seen any job sub script overriding this param.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml [hunk 5]
```diff
@@ -169,24 +167,24 @@ model:
     num_workers: 4
     pin_memory: true
     speech_offset: 30128
-    train_task: asr
+    train_task: all
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml:170]: `train_task` seems default as `tts` before. I understand either `tts` or `all` should work well if "Text to speech this" in question_in_manifest or "Phoneme TTS" in question_in_manifest.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml [hunk 6]
```diff
@@ -204,8 +202,8 @@ model:
         ascii_letter_case: "upper"
       german:
         _target_: nemo.collections.tts.g2p.models.i18n_ipa.IpaG2p
-        phoneme_dict: "/home/jasoli/gitrepos/NeMo/scripts/tts_dataset_files/de/de_nv240125.dict"
-        heteronyms: "/home/jasoli/gitrepos/NeMo/scripts/tts_dataset_files/de/de_nv230119.heteronym"
+        phoneme_dict: "scripts/tts_dataset_files/de/de_nv230119.dict"
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml:205]: any reason why we change to the previous version?

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py [hunk 7]
```diff
@@ -1416,14 +1416,14 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> A
...
-            for t in range(self.decoder_context_len, dec_input.shape[2] - 1):
+            for t in range(self.decoder_context_len + 1, dec_input.shape[2] - 1):
                 # Start at 0 if encoder context, else context_len
                 if t % 100 == 0:
                     logging.info("Timestep {}".format(t))
                 if t == end_inference_loop_at:
                     print("All ends detected")
                     break
-                if t == self.decoder_context_len:
+                if t == self.decoder_context_len + 1:
```
Comment (XuesongYang) [inline nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py:1426]: nit: sanity check, is `+1` a typo?

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml (hunk 6 above)
Comment (blisc) [reply examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml:205]: I don't think I pushed this new version to github. The only difference was that I removed `aneinander	ˌan1aɪnˈandɜ` as that was the only entry that had suprasegmentals and I didn't want to add a {1} token to my tokenizer. When we revisit multilingual, we should revisit this decision.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml (hunk 4 above)
Comment (blisc) [reply examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml:72]: I have trained with flash_attention. TBH, I haven't seen any performance speed impact nor any training / val loss impact. Since we add the attention prior, we only use flash attention in certain attention layers (self-attention in encoder, self-attention in decoder, cross-attention in decoder ONLY when we turn off the prior, don't use ctc loss, or don't want to return the attention probabilities). Cross-attention in decoder only uses flash attention during validation on iterations where we don't plot attention probabilities.

Comment (XuesongYang) [reply examples/tts/speechllm/conf/megatron_t5_speechlm_model.yaml:72]: i see. Previous recipes default as True, so from now on, we all start with False.

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py (hunk 7 above)
Comment (paarthneekhara) [reply nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py:1426]: I think this is because, we need an extra token after decoder context (which is a padding token on which loss is masked during training). I think there is a zero padding between context and answer, and during inference we need to pass that padding as input to correctly start predicting the answer.
